server/py/framework/utils/singletons/k8s.py

```diff
@@ -1132,19 +1132,147 @@ def _should_update_token_secret(
         self, k8s_secret: client.V1Secret, new_expiration: int
     ) -> bool:
         """
-        Determine if the secret should be updated based on tokenExpiration
+        Determine if the secret should be updated based on tokenExpiration.
+
+        :param k8s_secret: Existing Kubernetes secret.
+        :param new_expiration: Expiration timestamp of the new token.
+        :return: True if the secret should be updated, False otherwise.
         """
-        if not k8s_secret.data or "tokenExpiration" not in k8s_secret.data:
+        existing_exp = self._decode_secret_expiration(k8s_secret)
+
+        # If no expiration could be decoded, assume it needs an update
+        if existing_exp is None:
             return True
 
+        return new_expiration > existing_exp
+
+    def list_user_token_secrets(
+        self,
+        username: str,
+        namespace: typing.Optional[str] = None,
+    ) -> list[mlrun.common.schemas.SecretTokenInfo]:
+        """
+        List all offline token secrets for a given user.
+
+        :param username: The user whose tokens should be listed.
+        :param namespace: Kubernetes namespace where the secrets are stored.
+        :return: List of SecretTokenInfo objects, each containing the token name and expiration.
+        """
+        namespace = self.resolve_namespace(namespace)
+        labels = {
+            mlrun_constants.MLRunInternalLabels.user_token_secret_label_key: username
+        }
+
+        logger.debug(
+            "Listing user token secrets", username=username, namespace=namespace
+        )
+
+        k8s_secrets = self.list_secrets(namespace=namespace, labels=labels)
+
+        secret_tokens: list[mlrun.common.schemas.SecretTokenInfo] = []
+
+        for k8s_secret in k8s_secrets:
+            secret_name = k8s_secret.metadata.name
+
+            prefix = (
+                mlrun.mlconf.secret_stores.kubernetes.user_token_secret_name.format(
+                    username=username, token_name=""
+                )
+            )
+            if not secret_name.startswith(prefix):
+                logger.warning(
+                    "Skipping secret with unexpected name format",
+                    secret_name=secret_name,
+                )
+                continue
+
+            token_name = secret_name[len(prefix) :]
+
+            expiration = self._decode_secret_expiration(k8s_secret)
+            if expiration is None:
+                continue
+
+            secret_tokens.append(
+                mlrun.common.schemas.SecretTokenInfo(
+                    name=token_name,
+                    expiration=expiration,
+                )
+            )
```

Comment:
May be a helper method for converting a single k8s secret to a `mlrun.common.schemas.SecretTokenInfo`